xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,259 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.Internal;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Serves as a wrapper for {@link XdsClientImpl} used on the server side by {@link
+ * io.grpc.xds.internal.sds.XdsServerBuilder}.
+ */
+@Internal
+public final class XdsClientWrapperForServerSds {
+  private static final Logger logger =
+      Logger.getLogger(XdsClientWrapperForServerSds.class.getName());
+
+  private static final EventLoopGroupResource eventLoopGroupResource =
+      Epoll.isAvailable() ? new EventLoopGroupResource("GrpcServerXdsClient") : null;
+
+  private EnvoyServerProtoData.Listener curListener;
+  // TODO(sanjaypujare): implement shutting down XdsServer which will need xdsClient reference
+  @SuppressWarnings("unused")
+  @Nullable private XdsClient xdsClient;
+  private final int port;
+
+  /** Creates a new instance. */
+  public static XdsClientWrapperForServerSds newInstance(
+      int port, Bootstrapper bootstrapper, SynchronizationContext syncContext) throws IOException {
+    Bootstrapper.BootstrapInfo bootstrapInfo = bootstrapper.readBootstrap();
+    final List<Bootstrapper.ServerInfo> serverList = bootstrapInfo.getServers();
+    if (serverList.isEmpty()) {
+      throw new NoSuchElementException("No management server provided by bootstrap");
```

Comment:
I doubt the usage of `NoSuchElementException` here and `IllegalStateException` below. Error handling still looks a bit abrupt, basically it would panic immediately when encountering those cases. @creamsoup Does server side have any error propagation APIs?

Reply:
No these are caught by the caller where it will switch over to plaintext. As mentioned elsewhere there is no error propagation/error listener similar to what you have on the client side.

Reply:
netty installs `WriteBufferingAndExceptionHandler` which handles exception. but, this specific exception happens on `XdsServerBuilder#build`.

Reply:
> No these are caught by the caller where it will switch over to plaintext. As mentioned elsewhere there is no error propagation/error listener similar to what you have on the client side.

I still do not follow where these runtime exceptions are handled.

Reply:
https://github.com/grpc/grpc-java/pull/6838/files#diff-fe0356a9c9642d0f91961b377c24e5d3R265 line 265 in SdsProtocolNegotiators.java (in this commit). Just follow calls to `XdsClientWrapperForServerSds.newInstance`.

Reply:
That's catching the IOException. I am talking about `NoSuchElementException` here and `IllegalStateException` exception below. They are not gracefully handled. It panic the server indefinitely once getting into those situations.

Reply:
Good point, thanks for catching that. The `IllegalStateException` will go away since I will use `ScheduledThreadPoolExecutor` as we discussed in the other thread. And I will handle `NoSuchElementException` (or whatever is suitable type of exception for that case).